coderd/database/migrations/000034_linked_user_id.up.sql

```diff
@@ -0,0 +1,53 @@
+BEGIN;
+
+CREATE TABLE IF NOT EXISTS user_links (
+	user_id uuid NOT NULL,
+	login_type login_type NOT NULL,
+	linked_id text DEFAULT ''::text NOT NULL,
+	oauth_access_token text DEFAULT ''::text NOT NULL,
+	oauth_refresh_token text DEFAULT ''::text NOT NULL,
+	oauth_expiry timestamp with time zone DEFAULT '0001-01-01 00:00:00+00'::timestamp with time zone NOT NULL,
+	UNIQUE(user_id, login_type)
+);
+
+-- This migrates columns on api_keys to the new user_links table.
+-- It does this by finding all the API keys for each user, choosing
+-- the most recently updated for each one and then assigning its relevant
+-- values to the user_links table.
+-- A user should at most have a row for an OIDC account and a Github account.
+-- 'password' login types are ignored.
+
+INSERT INTO user_links 
+	( 
+		user_id,
+		login_type,
+		linked_id,
+		oauth_access_token,
+		oauth_refresh_token,
+		oauth_expiry
+	)
+SELECT 
+	keys.user_id, 
+	keys.login_type,
+	'',
+	keys.oauth_access_token,
+	keys.oauth_refresh_token,
+	keys.oauth_expiry 
+FROM 
+	( 
+		SELECT 
+			row_number() OVER (partition by user_id, login_type ORDER BY updated_at DESC) AS x, 
+			api_keys.* FROM api_keys
+	) as keys
+ WHERE x=1 AND keys.login_type != 'password';
```

Comment:
Did you just test this manually? I'm assuming so.

Reply:
Yup along with consulting @coadler. There's not really a great way to test this otherwise right now right?